sphinx/builders/__init__.py

```diff
@@ -75,15 +77,21 @@ class Builder:
     #: The builder supports data URIs or not.
     supported_data_uri_images = False
 
-    def __init__(self, app: "Sphinx") -> None:
+    def __init__(self, app: "Sphinx", env: BuildEnvironment = None) -> None:
```

Comment:
```suggestion
    def __init__(self, app: "Sphinx", env: Optional[BuildEnvironment] = None) -> None:
```

Reply:
Here I'm using `None` as a sentinel, I could have equally used `...`. I don't want to indicate to users that `None` is a valid type, which I believe using `Optional[]` syntax would do. Hence I'm -1 on both these suggestions, unless you've alternate ideas?

A

Reply:
that's not how this works. Setting a default value of `None` effectively coerces the type from `Type` to `Optional[Type]`. That is, the former desugars to the latter *implicitly*. The setting that controls this is called [`no-implicit-optional`](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-no-implicit-optional).

It's better to be explicit and use the `Optional`. If using 'strict' mypy checking, failing to use it becomes a hard error.

Reply:
I added `no_implicit_optional = True` and everything still passes -- is this what you meant?

A

Reply:
Yep, it should be an error with that config. Not sure why you're not seeing one, could be you're hitting the maximum number of errors, or it could be a false negative. In either case you should be able to convince yourself from the documentation that using the explicit `Optional` is strictly correct.

Reply:
ah i think i see the issue. you have this-

```ini
strict_optional = False
no_implicit_optional = True
```

to see these errors your would need

```ini
strict_optional = True
no_implicit_optional = True
```

if you really want to give yourself nightmares, try

```ini
strict = True
```